### ISSUE TITLE
can't identify what's wrong with the response 1406

### ISSUE THREAD

**Issue by jbfaden** [hapi-server/verifier-nodejs#10]
Can't figure out what's wrong with 

http://sarahandjeremy.net/pihapi/hapi/info?id=a_test_of_an_invalid_id_by_verifier-nodejs

where I get the error:

Warn: Want HTTP message to match 'HAPI' and '1406'' for clients who do not have access to response body for HTTP 400-level errors; Got: Not Found

Try http://tsds.org/verify-hapi?url=http://sarahandjeremy.net/pihapi/hapi, and it appears about 17 tests down (half-screen down).

**Comment by rweigel** [comment]
The verifier finds

HTTP/1.1
​ ​
404 Not Found


but wants something like

HTTP/1.1
​ ​
404

Not Found
HAPI 1406

This check was added for clients who are using a library that won't return
the body when the HTTP code is not 200. I don't recall which clients this
is for.

We never decided on the format of the recommend error message so the
verifier just looks for 1406 and HAPI in the status message string.

**Comment by jbfaden** [comment]
I thought we were supposed to respond with a hapi JSON, with a status area that says not found.  

Bring up http://sarahandjeremy.net/pihapi/hapi/info?id=a_test_of_an_invalid_id_by_verifier-nodejs in Chrome, you get:

{ "HAPI": "2.0", "status": { "code": 1406, "message": "error" } }

So should I have { ... "message": "HAPI 1406" } } ?  

Also, is it not possible to have a link to the javascript code that is raising the error?  I find it extremely helpful to be able to look at what the test is expecting...

**Comment by rweigel** [comment]
Your response body is valid and does not need to be modified. To avoid the
warning, you need to change the HTTP header message.

The code is looking at the first line of the raw HTTP header. It is usually
something like

HTTP/1.1 200 OK

HTTP/1.1 404 Not Found

HTTP/1.1 301 Moved Permanently

etc.  The text following the number is the HTTP status message.

This first line (what you see when you use curl -I http://...) can be
modified by a server. Your header did not throw an error because you had
HTTP/1.1 404, which is correct according to the spec. Your header triggered
a warning only.

The test is at
https://github.com/hapi-server/verifier-nodejs/blob/master/tests.js#L348

You got a warning because statusMessage (the OK, Not Found, Moved
Permanently) did not contain HAPI 1406.

**Comment by jbfaden** [comment]
I still don't see what the problem is...  Here's the output of curl -v:

~~~~~
spot5> curl -v  'http://jfaden.net/pihapi/hapi//info?id=a_test_of_an_invalid_id_by_verifier-nodejs'
* Hostname was NOT found in DNS cache
*   Trying 192.168.0.206...
* Connected to jfaden.net (192.168.0.206) port 80 (#0)
> GET /pihapi/hapi//info?id=a_test_of_an_invalid_id_by_verifier-nodejs HTTP/1.1
> User-Agent: curl/7.35.0
> Host: jfaden.net
> Accept: */*
> 
< HTTP/1.1 404 Not Found
< Date: Thu, 01 Feb 2018 17:07:32 GMT
* Server BaseHTTP/0.3 Python/2.7.9 is not blacklisted
< Server: BaseHTTP/0.3 Python/2.7.9
< Content-type: application/json
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET
< Access-Control-Allow-Headers: Content-Type
< Via: 1.1 localhost
< Transfer-Encoding: chunked
< 
{ "HAPI": "2.0", "status": { "code": 1406, "message": "Not Found" } }
* Connection #0 to host jfaden.net left intact

~
spot5> 
~~~~~
What should it be?  I know I shouldn't fuss about a warning, but I still don't understand what this test is telling me...  (And I almost get a perfect score on the test!)

**Comment by rweigel** [comment]
In a telecon discussion, we decided that it would be helpful if servers
said something like

HTTP/1.1 404 Not Found HAPI 1406

instead of what you show, which is

HTTP/1.1 404 Not Found

for 4xx and 5xx errors. The reason is that some client libraries won't let
the user see the body (the JSON you sent) if the header indicates an error.
Some clients assume that the body has no useful information and just return
an error and the header. We should probably modify the spec to indicate
exactly what the header should be for this optional server feature.

**Comment by jbfaden** [comment]
Okay, that makes sense now.  I don't know if I can control that, so maybe I'll just have to live with an A instead of A+.

**Comment by rweigel** [comment]
This is the same problem mentioned and fixed in #12
